### PR TITLE
Support globbing patterns for source-roots

### LINKED ISSUE #8290: Support globbing patterns for source-roots
### Current problem

It's quite inconvenient having to specify every source root for complex multi-package projects like `--source-roots src/package1,src/package2,...,src/packageN`

### Desired solution

For complex multi-package projects it would be nice to be able to specify source roots as `--source-roots src/*` instead of listing every one of them. IMHO, it's better to go with globbing patterns rather than with regexp patterns since those give better support for path-specific matching.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alexey-pelykh**
Add globbing pattern support for source-roots.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #8290

### CODE REVIEW

**Comment by ollie-iterators** [comment]
The documentation issue can be fixed by merging in changes from main.

**Comment by ollie-iterators** [comment]
https://github.com/PyCQA/pylint/pull/8286 fixed the issue with self-import

**Comment by DanielNoord** [comment]
I haven't read much of the discussion but was triggered about the comment about breaking changes in the accompanying issue. Note that the current implementation is definitely a breaking change as it changes the behaviour of an existing option. That's what we normally consider to be a breaking change.

**Comment by alexey-pelykh** [comment]
@DanielNoord it's not a breaking change since it's backwards compatible and on top the option in question hasn't been released yet as well

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas not a WIP anymore, yet there's a regression in test coverage since `_paths_csv_transformer` is no longer used. I'm not sure it's worth removing it really yet I'm open for options

Also, it would be much appreciated if this will go into the same release as the source-roots feature 🙏

**Comment by DudeNr33** [comment]
I really like the idea behind this.
However I share Pierre's opinion that this can get quite confusing for users now as this would only be available for the new `--source-roots` option, but not for others.

Users that learn glob patterns work for `--source-roots` might try to use the same for other options like `--ignore-paths` or simply when calling `pylint` (you could think of `pylint --recursive=y src/prefix**/*.py` or something similar) and spend quite some time trying to figure out why glob patterns work for one option but not for another.

A more holistic approach that takes into account the other options/arguments that accept paths would be better in my opinion.

**Comment by alexey-pelykh** [comment]
@DudeNr33 I agree and adding globbing support for the input sources is easy and non-breaking, yet `ignore-paths` e.g. are regexp-based right now so that would be breaking unless we add a new set of parameters. What would be your vision on overcoming that?

**Comment by DudeNr33** [comment]
> @DudeNr33 I agree and adding globbing support for the input sources is easy and non-breaking, yet `ignore-paths` e.g. are regexp-based right now so that would be breaking unless we add a new set of parameters. What would be your vision on overcoming that?

My take on this would be:

* support globbing where the current behaviour is not altered (i.e. anything that is not regex based probably)
* leave the regex based options as they are for v2.x
* change the default behavior from regex to globbing in v3.0, optionally adding support for people who really must use regex patterns by adding `--ignore-paths-regex` etc. 
* nice to have in v2.x: warn the user if an expression that is currently regex based cannot be evaluated as glob pattern, and make them aware that they must use regex pattern here

@DanielNoord your opinion?

**Comment by DanielNoord** [comment]
I think it is better to do globbing now for those settings where it is not a breaking change (this one). It just makes it easier for users to transition to `3.0` as they would need to update one less config option.

**Comment by alexey-pelykh** [comment]
So I'll add another PR to accomplish those goals so that all of it will appear in 2.17, right?

**Comment by alexey-pelykh** [comment]
@DudeNr33 @DanielNoord @Pierre-Sassoulas so, I've created #8310 and #8311 that can be addressed separately and would allow to unblock this PR.

**Comment by alexey-pelykh** [comment]
I guess I need some assistance with taming the CI as it fails randomly 🤔

File: doc/whatsnew/fragments/8290.feature [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Add globbing pattern support for source-roots.
```
Comment (DudeNr33) [inline doc/whatsnew/fragments/8290.feature:1]:
```suggestion
Add globbing pattern support for ``--source-roots``.
```

**Review by DudeNr33 — APPROVED** [review]
I re-ran the CI jobs and they passed now.
I only got one minor remark on the changelog entry, but that is not a must.

**Comment by alexey-pelykh** [comment]
Thanks, I've merged the suggestion

**Review by DudeNr33 — APPROVED** [review]
Thanks!

@Pierre-Sassoulas with the follow-up tasks for 3.0 created I think we can put this MR in the 2.17 release. I'm pinging you as you originally set the 3.0 milestone.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**